### PR TITLE
Begin replacement of hsTArray with std::vector

### PR DESCRIPTION
**Pull request by zrax**
This is far from complete, but it seemed like a decent start for several "easy" cases and a few more complicated ones...

### CODE REVIEW

File: Sources/Plasma/PubUtilLib/plDrawable/plDrawableGenerator.cpp [hunk 1]
```diff
@@ -628,20 +628,20 @@ plDrawableSpans     *plDrawableGenerator::GenerateAxesDrawable( hsGMaterial *mat
...
     plDrawableSpans     *drawable;
 
 
     /// Generate points
-    points.SetCount( 6 * 3 );
-    normals.SetCount( 6 * 3 );
-    colors.SetCount( 6 * 3 );
+    points.resize(6 * 3);
+    normals.resize(6 * 3);
+    colors.resize(6 * 3);
```
Comment (Hoikas) [inline Sources/Plasma/PubUtilLib/plDrawable/plDrawableGenerator.cpp:644]: Why not move these to the initializers?

File: Sources/Plasma/PubUtilLib/plDrawable/plGeoSpanDice.cpp [hunk 2]
```diff
@@ -193,12 +193,11 @@ plGeometrySpan* plGeoSpanDice::IExtractTris(plGeometrySpan* src, hsTArray<uint32
 {
     // First off, find out how many and which vers we're talking here.
     // Easiest way is while we're building the LUTs we'll want later anyway.
-    hsTArray<int16_t> fwdLUT;
-    fwdLUT.SetCount(src->fNumVerts);
-    memset(fwdLUT.AcquireArray(), -1, src->fNumVerts * sizeof(*fwdLUT.AcquireArray()));
+    std::vector<int16_t> fwdLUT;
+    fwdLUT.resize(src->fNumVerts);
+    memset(fwdLUT.data(), -1, src->fNumVerts * sizeof(*fwdLUT.data()));
```
Comment (Hoikas) [inline Sources/Plasma/PubUtilLib/plDrawable/plGeoSpanDice.cpp:198]: Why not?

```suggestion
    std::vector<int16_t> fwdLUT(src->fNumVerts, -1);
```

File: Sources/Plasma/PubUtilLib/plDrawable/plWaveSet7.cpp [hunk 3]
```diff
@@ -3564,15 +3564,15 @@ plDrawableSpans* plWaveSet7::ICreateGraphDrawable(plDrawableSpans* drawable, hsG
...
-    pos.SetCount(nVerts);
-    norm.SetCount(nVerts);
+    pos.resize(nVerts);
+    norm.resize(nVerts);
 
 #ifdef TEST_UVWS
-    hsTArray<hsPoint3> uvw; // Are we actually ever going to use these uvws?
-    uvw.SetCount(nVerts);
+    std::vector<hsPoint3> uvw; // Are we actually ever going to use these uvws?
+    uvw.resize(nVerts);
```
Comment (Hoikas) [inline Sources/Plasma/PubUtilLib/plDrawable/plWaveSet7.cpp:3575]: Why not allocate in the initializer?

File: Sources/Plasma/PubUtilLib/plDrawable/plWaveSet7.cpp [hunk 4]
```diff
@@ -3610,10 +3610,10 @@ plDrawableSpans* plWaveSet7::ICreateGraphDrawable(plDrawableSpans* drawable, hsG
 
     const int nTris = (nWid-1) * 2;
 
-    hsTArray<uint16_t> idxArr;
-    idxArr.SetCount(nTris * 3);
+    std::vector<uint16_t> idxArr;
+    idxArr.resize(nTris * 3);
```
Comment (Hoikas) [inline Sources/Plasma/PubUtilLib/plDrawable/plWaveSet7.cpp:3614]:
```suggestion
    std::vector<uint16_t> idxArr(nTris * 3);
```

File: Sources/Tools/MaxComponent/plClusterComponent.cpp [hunk 5]
```diff
@@ -786,14 +786,14 @@ void plClusterComponent::IFinishDoneNodes(INodeTab& doneNodes, plBox3Tab& fade,
...
-    hsTArray<hsRadixSortElem> scratchList;
+    std::vector<hsRadixSortElem> scratchList;
 
     int numClust = fCompPB->Count(kClusters);
 
     if( !numClust )
         return;
 
-    scratchList.SetCount(numClust);
+    scratchList.resize(numClust);
```
Comment (Hoikas) [inline Sources/Tools/MaxComponent/plClusterComponent.cpp:796]: Why not allocate in the initializer?

File: Sources/Plasma/FeatureLib/pfPython/cyParticleSys.cpp [hunk 6]
```diff
@@ -54,15 +54,15 @@ cyParticleSys::cyParticleSys(plKey sender, plKey recvr)
 }
 
 // setters
-void cyParticleSys::SetSender(plKey &sender)
+void cyParticleSys::SetSender(const plKey &sender)
 {
     fSender = sender;
```
[the remark this answers is not in the document]
Comment (zrax) [reply Sources/Plasma/FeatureLib/pfPython/cyParticleSys.cpp:59]: Many of these changes were made before I made `plKey` movable, so yeah, they should be revisited...

File: Sources/Plasma/CoreLib/_CoreLibPch.h [hunk 7]
```diff
@@ -49,6 +49,7 @@ You can contact Cyan Worlds, Inc. by email legal@cyan.com
 #include <functional>
 #include <limits.h>
 #include <memory>
+#include <vector>
 #include <stdarg.h>
 #include <stdint.h>
 #include <wchar.h>
```
[the remark this answers is not in the document]
Comment (zrax) [reply Sources/Plasma/CoreLib/_CoreLibPch.h:55]: Sure, and we should probably use the `cstdarg` etc. versions of those headers anyway...

File: Sources/Plasma/FeatureLib/pfSurface/plDistOpacityMod.cpp [hunk 8]
```diff
@@ -266,29 +260,20 @@ void plDistOpacityMod::ISetup()
...
-                    MatLayer* push = todo.Push();
-                    push->fMat = mat;
-                    push->fLay = lay;
-                }
+                auto it = std::find_if(todo.cbegin(), todo.cend(),
+                                       [lay](const MatLayer& matLayer) {
+                                           return lay == std::get<1>(matLayer);
+                                       });
+                if (it == todo.end())
+                    todo.emplace_back(mat, lay);
```
[the remark this answers is not in the document]
Comment (zrax) [reply Sources/Plasma/FeatureLib/pfSurface/plDistOpacityMod.cpp:268]: That code isn't the same...  It currently only adds the todo item only if the layer isn't found in the list; your rework would add it if any single item doesn't match.

**Review by Hoikas — APPROVED** [review]

**Pull request merged**